File: csharp/src/ITransformable.cs
using System.Linq;

namespace Vim.Math3d
{
    public interface ITransformable3D<TSelf>
    {
        TSelf Transform(Matrix4x4 mat);
    }

    public static class Transformable3D
    {
        public static Matrix4x4 Multiply(params Matrix4x4[] matrices)
            => matrices.Aggregate(Matrix4x4.Identity, (m1, m2) => m1 * m2);

        public static T Transform<T>(this ITransformable3D<T> self, params Matrix4x4[] matrices)
            => self.Transform(Multiply(matrices));

        public static T Translate<T>(this ITransformable3D<T> self, Vector3 offset)
            => self.Transform(Matrix4x4.CreateTranslation(offset));

        public static T Translate<T>(this ITransformable3D<T> self, float x, float y, float z)
            => self.Translate(new Vector3(x, y, z));

        public static T Rotate<T>(this ITransformable3D<T> self, Quaternion q)
            => self.Transform(Matrix4x4.CreateRotation(q));

        public static T Scale<T>(this ITransformable3D<T> self, float scale)
            => self.Scale(new Vector3(scale, scale, scale));

        public static T Scale<T>(this ITransformable3D<T> self, Vector3 scales)
            => self.Transform(Matrix4x4.CreateScale(scales));

        public static T Scale<T>(this ITransformable3D<T> self, float x, float y, float z)
            => self.Scale(new Vector3(x, y, z));

        public static T ScaleX<T>(this ITransformable3D<T> self, float x)
            => self.Scale(x, 0, 0);

        public static T ScaleY<T>(this ITransformable3D<T> self, float y)
            => self.Scale(0, y, 0);

        public static T ScaleZ<T>(this ITransformable3D<T> self, float z)
            => self.Scale(0, 0, z);

        public static T LookAt<T>(this ITransformable3D<T> self, Vector3 cameraPosition, Vector3 cameraTarget, Vector3 cameraUpVector)
            => self.Transform(Matrix4x4.CreateLookAt(cameraPosition, cameraTarget, cameraUpVector));

        public static T RotateAround<T>(this ITransformable3D<T> self, Vector3 axis, float angle)
            => self.Transform(Matrix4x4.CreateFromAxisAngle(axis, angle));

        public static T Rotate<T>(this ITransformable3D<T> self, float yaw, float pitch, float roll)
            => self.Transform(Matrix4x4.CreateFromYawPitchRoll(yaw, pitch, roll));

        public static T Reflect<T>(this ITransformable3D<T> self, Plane plane)
            => self.Transform(Matrix4x4.CreateReflection(plane));

        public static T RotateX<T>(this ITransformable3D<T> self, float angle)
            => self.RotateAround(Vector3.UnitX, angle);

        public static T RotateY<T>(this ITransformable3D<T> self, float angle)
            => self.RotateAround(Vector3.UnitY, angle);

        public static T RotateZ<T>(this ITransformable3D<T> self, float angle)
            => self.RotateAround(Vector3.UnitZ, angle);

        public static T TranslateRotateScale<T>(this ITransformable3D<T> self, Vector3 pos, Quaternion rot, Vector3 scale) where T : ITransformable3D<T>
            => self.Translate(pos).Rotate(rot).Scale(scale);
    }
}

File: dart/lib/src/itransformable.dart
part of '../vim_math3d.dart';

abstract class ITransformable3D<TSelf> {
  TSelf transform(Matrix4x4 mat);
}

extension Transformable3D<T> on ITransformable3D<T> {
  static Matrix4x4 multiply(Iterable<Matrix4x4> matrices) {
    var seed = Matrix4x4.identity;
    for (var m in matrices) {
      seed = seed * m;
    }
    return seed;
  }

  T transform(List<Matrix4x4> matrices) {
    return this.transform(multiply(matrices));
  }

  T translate(Vector3 offset) {
    return this.transform(Matrix4x4.translation(offset));
  }

  T translateXyz(
    double x,
    double y,
    double z,
  ) {
    return translate(Vector3(x, y, z));
  }

  T rotate(Quaternion q) {
    return this.transform(Matrix4x4.rotation(q));
  }

  T scale(double scale) {
    return scaleVector(Vector3(scale, scale, scale));
  }

  T scaleVector(Vector3 scales) {
    return this.transform(Matrix4x4.scaleVector3(scales));
  }

  T scaleXyz(double x, double y, double z) {
    return scaleVector(Vector3(x, y, z));
  }

  T scaleX(double x) {
    return scaleXyz(x, 0.0, 0.0);
  }

  T scaleY(double y) {
    return scaleXyz(0.0, y, 0.0);
  }

  T scaleZ(double z) {
    return scaleXyz(0.0, 0.0, z);
  }

  T lookAt(
    Vector3 cameraPosition,
    Vector3 cameraTarget,
    Vector3 cameraUpVector,
  ) {
    return this.transform(
      Matrix4x4.lookAt(
        cameraPosition,
        cameraTarget,
        cameraUpVector,
      ),
    );
  }

  T rotateAround(Vector3 axis, double angle) {
    return this.transform(Matrix4x4.fromAxisAndAngle(axis, angle));
  }

  T rotateYpr(double yaw, double pitch, double roll) {
    return this.transform(Matrix4x4.fromYawPitchRoll(yaw, pitch, roll));
  }

  T reflect(Plane plane) {
    return this.transform(Matrix4x4.reflection(plane));
  }

  T rotateX(double angle) {
    return rotateAround(Vector3.unitX, angle);
  }

  T rotateY(double angle) {
    return rotateAround(Vector3.unitY, angle);
  }

  T rotateZ(double angle) {
    return rotateAround(Vector3.unitZ, angle);
  }
}

extension SelfTransformable3D<T> on ITransformable3D<ITransformable3D<T>> {
  ITransformable3D<T> translateRotateScale(
    Vector3 pos,
    Quaternion rot,
    Vector3 scale,
  ) {
    return translate(pos)
      ..rotate(rot)
      ..scaleVector(scale);
  }
}
